### ISSUE TITLE
Some questions about changing policies and observations

### ISSUE THREAD

**Issue by lucalazzaroni** [eleurent/highway-env#147]
Hi, I tried to run and make some changes to the "highway-v0" environment (i.e. no right overtake, safety distance and more...). I now have a question about training. At the moment the model structure is as follows:
```
model = DQN('MlpPolicy', env, gamma=0.8, learning_rate=5e-4, buffer_size=50000, exploration_fraction=0.1,
            exploration_final_eps=0.5, exploration_initial_eps=1.0, batch_size=32, double_q=True,
            target_network_update_freq=50, prioritized_replay=True, verbose=1, tensorboard_log="./dqn_two_lane_tensorboard/")
```
and observation type is Kinematics.
Results, after training sessions of 300000 steps are fluctuating, also adding layers to Mlp (64, 64, 64, 32, 20) which seems not to add anything to the standard Mlp. So I tried to use Grayscale observation and CnnPolicy, to see if there would be a performance improvement.
Here is the code:
```
model = DQN('CnnPolicy', env, gamma=0.8, learning_rate=5e-4, buffer_size=50000, exploration_fraction=0.1,
            exploration_final_eps=0.5, exploration_initial_eps=1.0, batch_size=32, double_q=True,
            target_network_update_freq=50, prioritized_replay=True, verbose=1, tensorboard_log="./dqn_two_lane_tensorboard/")
```
```
"offscreen_rendering": True,
"observation": {
    "type": "GrayscaleObservation",
    "weights": [0.2989, 0.5870, 0.1140],  # weights for RGB conversion
    "stack_size": 4,
    "observation_shape": (screen_width, screen_height)
},
"screen_width": screen_width,
"screen_height": screen_height,
"scaling": 1.75,
"policy_frequency": 2,
```
The training starts with no errors, but after some steps (around 4000) it crashes due to occupation of all the RAM memory. I tried to reduce batch size (up to 16) and screen width and height (up to 84x84 which is really small) but it doesn't change anything. 

My PC specs are the following:
GPU model: NVIDIA Quadro RTX 4000
CUDA version: 10.1
RAM: 32 GB

My question is if there is something I'm missing that causes the RAM saturation and, mostly, if using Cnn + Grayscale observation would actually result in a performance improvement or if it's a waste of time.
Thanks in advance for your help

**Comment by eleurent** [comment]
Hi @lucalazzaroni,

> if using Cnn + Grayscale observation would actually result in a performance improvement or if it's a waste of time.

I expect that using a CNN model + Image observation should indeed result in a better performance than the MLP model + Kinematics observation pair. I argued in [this paper](https://arxiv.org/abs/1911.12250) that the MLP model works quite badly with Kinematics observation probably because its output depends on the order in which vehicles are described. Thus, permutation-invariant architectures such as self-attention layers typically learn faster. And the image observation does not suffer from this limitation.

> The training starts with no errors, but after some steps (around 4000) it crashes due to occupation of all the RAM memory. I tried to reduce batch size (up to 16) and screen width and height (up to 84x84 which is really small) but it doesn't change anything.

I successfully trained CNN policies with the GrayscaleObservation on intersection-v0 a few months ago, and at the time I did not run into any RAM memory overflow error... I'll try it again on highway-v0 and see if I can reproduce the issue and spot a memory leak.

**Comment by eleurent** [comment]
I just launched a training run with a buffer size of 40000, image size of 128x64x4, and the buffer takes 12.6 Gb of RAM memory, out of my 32 Gb available. I'll see if it increases with time.

I also changed the image dtype from float32 to uint8 to save some space, and will push these changes asap.

**Comment by lucalazzaroni** [comment]
I just tried a training session on Colab to see if it's my PC fault. I set batch size to 8, image size 128x64 and buffer size 40000. I'm currently at 18000 steps and RAM usage is around 3 GB. I'll let you know if it increases.
For now, thanks for the support

**Comment by lucalazzaroni** [comment]
Such a strange behavior. RAM usage remained around 3 GB until 500 episodes (23415 steps). Then it rapidly increased to 100%, performance dropped as a consequence and after some minutes Colab runtime crashed for memory overflow error.

**Comment by eleurent** [comment]
Could it be due to the learning_starts: int = 50000 parameter, which could kick in after ~500 episodes of length ~10 and cause the memory error only then?

**Comment by lucalazzaroni** [comment]
> Could it be due to the learning_starts: int = 50000 parameter, which could kick in after ~500 episodes of length ~10 and cause the memory error only then?

How can I check it?

**Comment by eleurent** [comment]
Maybe try to set this parameter to 0 and see if you get the error right away.

**Comment by eleurent** [comment]
Update: after 200k steps, my RAM is at 15.8/32Gb and the learning curve seems reasonable:
![image](https://user-images.githubusercontent.com/1706935/111881222-cad26400-89af-11eb-9e54-55fba70f0abf.png)

**Comment by lucalazzaroni** [comment]
Thank you for the update. I tried setting learning_start to 0 as you suggested (on Colab since I can't access the PC until monday) and now I can see that RAM occupancy increases normally. The problem is that the overflow occurs at only ~30000 steps making proper training impossible.
Here is the code updated:
```
model = DQN('CnnPolicy', env, gamma=0.8, learning_rate=5e-4, buffer_size=40000, exploration_fraction=0.1,
            exploration_final_eps=0.5, exploration_initial_eps=1.0, batch_size=8, double_q=True,
            target_network_update_freq=50, prioritized_replay=True, verbose=1, learning_starts=int(0), tensorboard_log="./dqn_two_lane_tensorboard/")
```

Do you have any suggestions to avoid this kind of problem?
Thank you again for your time

**Comment by eleurent** [comment]
I could not reproduce the memory leak issue... On my computer, the training finished successfully at 400k steps, and memory usage stayed in the range 12.5Gb - 15.5 Gb.
Which version of stable baselines are you using? I use stable-baseline3 (the pytorch version), which apparently allocates the full buffer memory during the 1st episode rather than progressively with time as you observed. This seems better, as it allows to better estimate the amount of memory required by the observation shape and buffer size.

After those 400k episodes, DQN reached an average reward of about 30 (with my env configuration), which corresponds to the following behavior :
(2 episodes, in the observation resolution of 128x64, sorry about that...)
https://user-images.githubusercontent.com/1706935/111888769-c9b92b00-89df-11eb-9834-e8042627bb90.mp4

**Comment by lucalazzaroni** [comment]
I was using stable-baselines 2.10, today I tried to switch to stable-baselines3 as you suggested but I'm having issues with the observation type which does not match the type required from sb3. The error is as follows: `You should use NatureCNN only with images not with Box(0.0, 1.0, (128, 64, 4), float32)`. I run the env checker and in fact it returns 2 warnings about it: 

`It seems that your observation is an image but the 'dtype' of your observation_space is not 'np.uint8'. If your observation is not an image, we recommend you to flatten the observation to have only a 1D vector`

`It seems that your observation space is an image but the upper and lower bounds are not in [0, 255]. Because the CNN policy normalize automatically the observation you may encounter issue if the values are not in that range.`

I'm going to wait for your changes to the observation and the training script which I think will solve the problem. 
Thank you.

**Comment by eleurent** [comment]
Yes indeed, I run into that issue as well (the channel convention was WxHxC instead of CxWxH as required by sb3) and fixed it. Will push very soon, I'm finishing the last changes (having two separate viewers for env rendering and image observation)

**Comment by lucalazzaroni** [comment]
It works with no memory leak! I am really grateful to you for the help you have given me. You have been very helpful, thank you so much. Can close the issue.
